File: 1679-max-number-of-k-sum-pairs/1679-max-number-of-k-sum-pairs.go
func maxOperations(nums []int, k int) int {
    sort.Ints(nums)
    
    ops := 0
    left, right := 0, len(nums) - 1
    
    for left < right {
        lNum := nums[left]
        rNum := nums[right]
        if lNum + rNum == k {
            ops++
            left++
            right--
        } else if lNum + rNum < k {
            left++
        } else {
            right--
        }
    }
    return ops
}


/*
// Map Solution
func maxOperations(nums []int, k int) int {
    nCount := make(map[int]int)
    for _, n := range nums {
        nCount[n]++
    }
    
    ops := 0
    counted := make(map[int]bool)
    for n, count := range nCount {
        if !counted[n] {
            if n != (k-n) {
                ops += min(count, nCount[k-n])
            } else {
                ops += count / 2
            }
            counted[n] = true
            counted[k-n] = true
        }
    }
    return ops
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
*/

/*
// Map Solution (suboptimal)
func maxOperations(nums []int, k int) int {
    ops := 0
    for len(nums) != 0 {
        found := false
        comps := make(map[int]int)
        for i, n := range nums {
            if v, ok := comps[k-n]; ok {
                nums = append(nums[:i], nums[i+1:]...)
                nums = append(nums[:v], nums[v+1:]...)
                found = true
                ops++
                break;
            }
            comps[n] = i
        }
        if !found {
            break
        }
    }
    return ops
}
*/

File: 1679-max-number-of-k-sum-pairs/1679-max-number-of-k-sum-pairs.c
int partition(int *nums, int left, int right) {
    int p_idx, p, temp;
    p_idx = left;
    p = nums[left++];
    
    for (int i = left; i < right; i++) {
        if (nums[i] < p) {
            temp = nums[i];
            nums[i] = nums[left];
            nums[left++] = temp;
        }
    }
    nums[p_idx] = nums[left-1];
    nums[left-1] = p;
    return left-1;
}

void quicksort(int *nums, int left, int right) {
    int p;
    
    if (left >= right)
        return;
    
    p = partition(nums, left, right);
    quicksort(nums, left, p);
    quicksort(nums, p+1, right);
}


int compare(void const *a, void const *b) {
    return *(int *)a - *(int *)b;
}

int maxOperations(int* nums, int numsSize, int k){
    int left, right, lnum, rnum, ops;
    
    qsort(nums, numsSize, sizeof(int), compare);
    
    left = 0;
    right = numsSize - 1;
    ops = 0;
    
    while (left < right) {
        lnum = nums[left];
        rnum = nums[right];
        if (lnum + rnum == k) {
            left++;
            right--;
            ops++;
        } else if (lnum + rnum < k)
            left++;
        else
            right--;
    }
    
    return ops;
}